src/test/java/com/puppycrawl/tools/checkstyle/JavaAstVisitorTest.java

```diff
@@ -190,4 +204,34 @@ public void testNullSelfInAddLastSibling() throws Exception {
                 .that(addLastSibling.invoke(JavaAstVisitor.class, null, null))
                 .isNull();
     }
+
+    @Test
+    public void testNoStackOverflowOnDeepStringConcat() throws Exception {
+        final File file =
+                new File(getPath("InputJavaAstVisitorNoStackOverflowOnDeepStringConcat.java"));
+        final FileText fileText = new FileText(file, StandardCharsets.UTF_8.name());
+        final FileContents contents = new FileContents(fileText);
+
+        final String fullText = contents.getText().getFullText().toString();
+        final CharStream codePointCharStream = CharStreams.fromString(fullText);
+        final JavaLanguageLexer lexer = new JavaLanguageLexer(codePointCharStream, true);
+        lexer.setCommentListener(contents);
+
+        final CommonTokenStream tokenStream = new CommonTokenStream(lexer);
+        final JavaLanguageParser parser = new JavaLanguageParser(tokenStream);
+
+        final JavaLanguageParser.CompilationUnitContext compilationUnit = parser.compilationUnit();
```

Comment:
Why aren't we using `JavaParser.parseFile`? Why did this change so much in the move?

Reply:
Because we do not want to clear the DFA when using using the limited resources hack. Consider:

We created the iterative expression AST building to avoid stackoverflow in `JavaAstVisitor#visitBinOp`. In actual generated parser, we avoid stackoverflow thanks to the left recursive expression rule (avoids unnecessary recursive calls to hierarchical expression production rules). But, upon clearing the DFA, ANTLR itself has no such optimization. This causes the intermittent stackoverflow reported in https://github.com/checkstyle/checkstyle/pull/10999#issuecomment-1001256625.

So, I basically rewrote `JavaParser#parse` here, using parser with no DFA clearing (single arg constructor).

Reply:
Please add a comment or something somewhere connected to this test that we can't use JavaParser here and why.

Reply:
Done

Reply:
Unresolved this, please see comments at https://github.com/checkstyle/checkstyle/pull/10999#discussion_r782964262 where this is fixed.

Reply:
I am not sure how this is unresolved because the comment was still added to the test.